ServiceWorkerRegistration data is not deleted correctly

szewai opened the pull request:
#### 8e9f2eb04c2e6dcb35850e785b6626545838af95
<pre>
ServiceWorkerRegistration data is not deleted correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=260163">https://bugs.webkit.org/show_bug.cgi?id=260163</a>
rdar://112092827

Reviewed by Brent Fulgham.

bindText() returns error code and 0 (SQLITE_OK) means the operation succeeds, so updateRegistrations should not return
early.

API Test: WKWebsiteDataStore.RemoveServiceWorkerDataByOrigin

* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
(WebCore::SWRegistrationDatabase::updateRegistrations):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: <a href="https://commits.webkit.org/266882@main">https://commits.webkit.org/266882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ba01c08490e5dfdf5ae14bdd1b8dbe69d1af953

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15335 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16786 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/14141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15435 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16768 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12772 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17515 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12963 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13558 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20522 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14041 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16964 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14287 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12089 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13567 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3620 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17903 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14128 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
```diff
@@ -433,7 +433,7 @@ std::optional<Vector<ServiceWorkerScripts>> SWRegistrationDatabase::updateRegist
     for (auto& registration : registrationsToDelete) {
         auto statement = cachedStatement(StatementType::DeleteRecord);
-        if (!statement || !statement->bindText(1, registration.toDatabaseKey()) || statement->step() != SQLITE_DONE) {
+        if (!statement || statement->bindText(1, registration.toDatabaseKey()) != SQLITE_OK || statement->step() != SQLITE_DONE) {
```
I don't understand how `!statement->bindText(...)` gives a different answer than `statement->bindText(...) != SQLITE_OK`, since `SQLITE_OK` is defined as `0`. Is it possible that we want to be checking for multiple things like maybe `SQLITE_DONE` or something?

brentfulgham replied:
On the other hand, all other instances of `bindText(...)` use this specific equality check, so this change is good if only as a way of making the code more consistent.

brentfulgham replied:
Sihui replied to me privately to explain that I am not doing math properly, and that the first case is true when `bindText` returns zero, and the new case is true when `bindText` does NOT return zero, which is actually the behavior we want.

szewai committed:
ServiceWorkerRegistration data is not deleted correctly
https://bugs.webkit.org/show_bug.cgi?id=260163
rdar://112092827

Reviewed by Brent Fulgham.

bindText() returns error code and 0 (SQLITE_OK) means the operation succeeds, so updateRegistrations should not return
early.

API Test: WKWebsiteDataStore.RemoveServiceWorkerDataByOrigin

* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
(WebCore::SWRegistrationDatabase::updateRegistrations):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:

Canonical link: https://commits.webkit.org/266882@main
